haddock-api/src/Haddock/Types.hs

```diff
@@ -640,8 +653,49 @@ data SinceQual
 -- A monad which collects error messages, locally defined to avoid a dep on mtl
 
 
-type ErrMsg = String
-type ErrMsgM = Writer [ErrMsg]
+type ErrMsg = Builder
+
+errMsgFromString :: String -> ErrMsg
+errMsgFromString = fromString
+
+errMsgToString :: ErrMsg -> String
+errMsgToString = Text.unpack . Text.decodeUtf8 . BSL.toStrict . toLazyByteString
+
+errMsgUnlines :: [ErrMsg] -> ErrMsg
+errMsgUnlines = List.foldl' (\acc x -> acc <> (x <> charUtf8 '\n')) mempty
+
+class Monad m => ReportErrorMessage m where
+    reportErrorMessage :: Builder -> m ()
+
+instance ReportErrorMessage m => ReportErrorMessage (ReaderT r m) where
+    reportErrorMessage = lift . reportErrorMessage
+
+#if !MIN_VERSION_mtl(2,3,0)
+-- | @since 2.3
+instance (Monoid w, Monad m) => MonadWriter w (CPS.WriterT w m) where
+    writer = CPS.writer
+    tell   = CPS.tell
+    listen = CPS.listen
+    pass   = CPS.pass
+#endif
+
+instance Monad m => ReportErrorMessage (WriterT ErrorMessages m) where
+    reportErrorMessage = tell . singleMessage
+
+newtype ErrMsgM a = ErrMsgM { unErrMsgM :: Writer ErrorMessages a }
+    deriving newtype (Functor, Applicative, Monad, ReportErrorMessage)
+
+newtype ErrorMessages = ErrorMessages { unErrorMessages :: DList Builder }
+    deriving newtype (Semigroup, Monoid)
+
+runErrMsgM :: ErrMsgM a -> (a, ErrorMessages)
+runErrMsgM = runWriter . unErrMsgM
+
+singleMessage :: Builder -> ErrorMessages
+singleMessage = ErrorMessages . pure
+
+errorMessagesToList :: ErrorMessages -> [Builder]
+errorMessagesToList = DList.toList . unErrorMessages
```

Comment:
You did a great job abstracting the error message business here. I think it'd be neat and tidy still if you'd inline the DList ala

```
newtype ErrorMessages = ErrorMessages { unErrorMessages :: [Builder] -> [Builder] }
    deriving newtype (Semigroup, Monoid)

singleMessage message = ErrorMessages . (message :)

errorMessagesToList (ErrorMessages messages) = messages []
```

reducing the amount of new code tremendously. WDYT?

Reply:
Good call! Ends up being just as fast, too. :)